src/serialization/v2_serializer.rs

```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+fn max_encoded_size<T: Counter>(h: &Histogram<T>) -> Option<usize> {
+    counts_array_max_encoded_size(h.index_for(h.max()) + 1)
+        .and_then(|x| x.checked_add(V2_HEADER_SIZE))
+}
+
+pub fn counts_array_max_encoded_size(length: usize) -> Option<usize> {
+    // LEB128-64b9B uses at most 9 bytes
+    // Won't overflow (except sometimes on 16 bit systems) because largest possible counts
+    // len is 47 buckets, each with 2^17 half count, for a total of 6e6. This product will
+    // therefore be about 5e7 (50 million) at most.
+    length.checked_mul(9)
+}
+
+/// Encode counts array into slice.
+/// The slice must be at least 9 * the number of counts that will be encoded.
+pub fn encode_counts<T: Counter>(h: &Histogram<T>, buf: &mut [u8]) -> Result<usize, SerializeError> {
+    let index_limit = h.index_for(h.max()) + 1;
+    let mut index = 0;
+    let mut bytes_written = 0;
+
+    assert!(index_limit <= h.counts.len());
+
+    while index < index_limit {
+        // index is inside h.counts because of the assert above
+        let count = unsafe { *(h.counts.get_unchecked(index)) };
+        index += 1;
+
+        // Non-negative values are positive counts or 1 zero, negative values are repeated zeros.
+
+        let mut zero_count = 0;
+        if count == T::zero() {
+            zero_count = 1;
+
+            // index is inside h.counts because of the assert above
+            while (index < index_limit) && (unsafe { *(h.counts.get_unchecked(index)) } == T::zero()) {
+                zero_count += 1;
+                index += 1;
+            }
+        }
+
+        let count_or_zeros: i64 = if zero_count > 1 {
+            // zero count can be at most the entire counts array, which is at most 2^24, so will fit.
+            -zero_count
+        } else {
+            count.to_i64().ok_or(SerializeError::CountNotSerializable)?
+        };
+
+        let zz = zig_zag_encode(count_or_zeros);
+
+        bytes_written += varint_write(zz, &mut buf[bytes_written..]);
+    }
+
+    Ok(bytes_written)
+}
+
+/// Write a number as a LEB128-64b9B little endian base 128 varint to buf. This is not
+/// quite the same as Protobuf's LEB128 as it encodes 64 bit values in a max of 9 bytes, not 10.
+/// The first 8 7-bit chunks are encoded normally (up through the first 7 bytes of input). The last
+/// byte is added to the buf as-is. This limits the input to 8 bytes, but that's all we need.
+/// Returns the number of bytes written (in [1, 9]).
+#[inline]
+pub fn varint_write(input: u64, buf: &mut [u8]) -> usize {
+    // The loop is unrolled because the special case is awkward to express in a loop, and it
+    // probably makes the branch predictor happier to do it this way.
+    // This way about twice as fast as the other "obvious" approach: a sequence of `if`s to detect
+    // size directly with each branch encoding that number completely and returning.
+
+    // TODO try bitwise and instead of shift
```

Comment:
`AND` is slower. Experimenting with hand-inlining all the helper method calls shows that the compiler already does a great job of hoisting out the common sub-expressions, and just the fact that `AND` is a different calculation than the shift (which eventually gets re-used in later branches) means that it ends up worse. I looked up those instructions in the latency tables and for a register and immediate, both `and` and `shr` have latency 1, so this result isn't too surprising. If the shifts weren't already being used, it should have been equivalent.